codex/stores/typeddatastore.nim

```diff
@@ -0,0 +1,109 @@
+import pkg/upraises
+
+push: {.upraises: [].}
+
+import std/sugar
+
+import pkg/chronos
+import pkg/chronos/futures
+import pkg/datastore
+import pkg/questionable
+import pkg/questionable/results
+import pkg/stew/results
+
+import ../utils
+import ../utils/genericcoders
+
+type
+  ModifyT*[T] = proc(v: ?T): Future[?T] {.upraises: [CatchableError], gcsafe, closure.}
+  ModifyTGetU*[T, U] = proc(v: ?T): Future[(?T, U)] {.upraises: [CatchableError], gcsafe, closure.}
+
+  BytesTuple = (?seq[byte], seq[byte])
+
+  KeyVal[T] = (?Key, ?!T)
+  ResIter[T] =  Future[KeyVal[T]]
+
+
+proc putT*[T](self: Datastore, key: Key, t: T): Future[?!void] {.async.} =
+  await self.put(key, t.encode)
+
+proc getT*[T](self: Datastore, key: Key): Future[?!T] {.async.} =
+  without bytes =? await self.get(key), errx:
+    return failure(errx)
+
+  T.decode(bytes)
+
+proc queryT*[T](self: Datastore, query: Query): Future[?!Iter[Future[KeyVal[T]]]] {.async.} =
```

Comment:
I agree that type signature is quite long and looks complicated. I think it's a good idea to make it less complicated by adding API options to handle encoding errors and missing value errors by the API impl, so basically the signature could be then shortened to `Future[?!AsyncIter[(Key, T)]]` which I think looks more friendly. Wdyt?

Reply:
> Wdyt?

My honest opinion is that this entire module is unnecessary and should be removed, because, as I explained several times in other places, the purpose of this module is to fetch strongly typed data from a datastore via decoding, which can be accomplished in a much more simple, readable and maintainable way using a pattern like the one I outlined in the `reservations` module.

Reply:
>using a pattern like the one I outlined in the reservations module

What's wrong with taking this pattern out to a separate module that can be imported and used in other places?